src/agent/Agent.cpp

```diff
@@ -0,0 +1,222 @@
+    tickless_ = next_discovery_;
+}
+
+
+void Agent::process_router() {
+    zmsg_t *reply = zmsg_recv(router_);
+
+    // frame 0 is server that replied
+    char *endpoint = zmsg_popstr(reply);
+    Server *server = (Server *) zhash_lookup(servers_, endpoint);
+    free(endpoint);
+
+    if (server == NULL) return;
+
+    server->alive_ = true;
+    server->refreshTimers();
+
+    // frame 1 is a control frame
+    char *control = zmsg_popstr(reply);
+
+    if (streq(control, "TASK")) {
+        char *status = zmsg_popstr(reply);
+        if (streq(status, "OK")) {
+            zmsg_destroy(&request_);
+            request_ = NULL;
+            unlock_servers();
+        } else if (streq(status, "RESULT") || streq(status, "ERROR")) {
+            zmsg_pushstr(reply, status);
+            zmsg_send(&reply, pipe_);
+            zmsg_destroy(&request_);
+            request_ = NULL;
+            unlock_servers();
+        } else if (streq(status, "BUSY")) {
+            server->busy_ = true;
+        }
+    } else {
+        zmsg_destroy (&reply);
+    }
+}
+
+void Agent::update_ping_time() {
+    Server* item = (Server*) zhash_first(servers_);
+    while(item) {
+        if (tickless_ > item->ping_at_) {
+            tickless_ = item->ping_at_;
+        }
+        item = (Server*) zhash_next(servers_);
+    }
+}
+
+// find server to send request to, remove any expired servers
+void Agent::process_request() {
+    if (!request_) return;
+
+    if (zclock_mono() >= expires_) {
+        //  Request expired, kill it
+        zmsg_t* expired = zmsg_new();
+        zmsg_pushstr(expired, "EXPIRED");
+        zmsg_pushstr(expired, "INTERNAL_ERROR");
+        zmsg_send(&expired, pipe_);
+        zmsg_destroy(&request_);
+        request_ = NULL;
+        unlock_servers();
+        return;
+    }
+
+    // TODO: packing/unpacking needs improvement or a new interface
+    // unpack message to get task_name, pack it back after
+    char* protocol_name = zmsg_popstr(request_);
+    char* task_name = zmsg_popstr(request_);
+    std::string task_name_str(task_name);
+    zmsg_pushstr(request_, task_name);
+    zmsg_pushstr(request_, protocol_name);
+    free(task_name);
+    free(protocol_name);
+
+    Server* server = (Server*) zhash_first(servers_);
+    while(server) {
+        // TODO: separate server expire check and request sending
+        if (zclock_mono () >= server->expires_) {
+            server->alive_ = false;
+        } else if (!server->busy_ && 
+                (server->supported_tasks_.find(task_name_str) != server->supported_tasks_.end())) {
+            // if server is not busy and supports the task, send a request
+            zmsg_t *request = zmsg_dup(request_);
+            zmsg_pushstr(request, server->endpoint_.c_str());
+            zmsg_send(&request, router_);
+            server->busy_ = true;
+            break;
+        }
+        server = (Server*) zhash_next(servers_);
```

Comment:
what does zhash_t do? is it possible that this is an std::map<std::string, server*>? or rather a std::map<std::string, server> (by value)?

Reply:
`zhash` stores (unique key)/value inside. This is very similar behavior to std::map. `zhash` was used in the guide that I was using as a base (http://zguide.zeromq.org/page:all#toc112). I was thinking of replacing it with `std::map` but at the end decide to leave due to the deadline. After the grading/presentation, I plan to replace it. It will also make function `s_server_free` not needed.